File: chat/web/src/rules.ts
/* Permission rules and history (parity round 2 B): the pure parts of the
   workspace panel's Permissions section, the card's "Allow always" scope
   and the chat menu's history list. The service (chats/rules.go) is the
   matcher and the record; this file checks a pattern before it is sent,
   words rules and decisions for people, and orders the history. */
import type { Actor, PermissionEvent, Rule } from "./types";

export const RULE_KINDS: { kind: Rule["kind"]; hint: string }[] = [
  { kind: "allow", hint: "Answered without asking in ask mode" },
  { kind: "deny", hint: "Refused without asking in every mode" },
  { kind: "ask", hint: "A card even in auto mode" },
];

export const RULE_EXAMPLES = [
  "Bash(git *)",
  "Bash(npm test)",
  "Edit(src/**)",
  "Read",
  "WebFetch(domain:example.com)",
  "mcp__warden__*",
];

/* Tools whose rules take a parenthesised argument (a command, a path
   glob, a domain); every other tool is written alone. */
const ARGUMENT_TOOLS = new Set([
  "Bash",
  "Edit",
  "Write",
  "MultiEdit",
  "NotebookEdit",
  "Read",
  "Glob",
  "Grep",
  "LS",
  "WebFetch",
]);

/* Why a pattern cannot be a rule, or "" when it can — the service checks
   again; this saves a round trip and words the same problems. */
export function ruleError(pattern: string): string {
  const p = pattern.trim();
  if (!p) return "Write a pattern, such as Bash(git *) or Edit(src/**).";
  if (p.length > 512 || /[\n\r]/.test(p)) return "One line, up to 512 characters.";
  const open = p.indexOf("(");
  const tool = open < 0 ? p : p.slice(0, open);
  if (!tool) return "Start with the tool's name: Bash, Edit, Read, WebFetch, mcp__server__tool.";
  if (!/^[A-Za-z0-9_*-]+$/.test(tool)) return `“${tool}” is not a tool name.`;
  if (open < 0) return "";
  if (!p.endsWith(")")) return "Missing the closing parenthesis.";
  const spec = p.slice(open + 1, -1).trim();
  if (!spec) return `Empty parentheses; write ${tool} alone.`;
  if (!ARGUMENT_TOOLS.has(tool)) return `${tool} rules take no argument; write ${tool} alone.`;
  if (tool === "WebFetch" && !/^domain:.+/.test(spec))
    return "WebFetch rules take domain:HOST, such as WebFetch(domain:example.com).";
  return "";
}

/* A person's name, email or principal. */
export function actorLabel(by?: Actor | null): string {
  if (!by) return "";
  if (by.name) return by.name;
  if (by.email) return by.email;
  if (by.principalID === "owner") return "the owner";
  return by.principalID;
}

/* Where a rule came from, for the list: the editor or an "Allow always"
   answer (in which chat), and who added it. */
export function ruleOrigin(
  rule: Rule,
  chats: { id: string; title: string }[] = [],
): string {
  let out = rule.origin === "always" ? "Allow always" : "from the editor";
  if (rule.origin === "always" && rule.chatID) {
    const chat = chats.find((c) => c.id === rule.chatID);
    if (chat) out += ` in “${chat.title}”`;
  }
  const by = actorLabel(rule.by);
  if (by) out += ` by ${by}`;
  return out;
}

/* What a kind does, one line, for a title. */
export function kindHint(kind: string): string {
  return RULE_KINDS.find((k) => k.kind === kind)?.hint ?? "";
}

/* Who or what decided an event, for the history list. */
export function decidedBy(ev: PermissionEvent): string {
  switch (ev.how) {
    case "auto":
      return "auto mode";
    case "rule":
      return ev.rule
        ? `${ev.scope ?? "chat"} rule ${ev.rule.kind} ${ev.rule.pattern}`
        : "a rule";
  }
  let by = actorLabel(ev.by) || "a card";
  if (ev.rule) by += `, always for the ${ev.scope ?? "chat"} (${ev.rule.pattern})`;
  if (ev.message) by += `: ${ev.message}`;
  return by;
}

/* The history newest first. */
export function newestFirst(events: PermissionEvent[]): PermissionEvent[] {
  return [...events].sort((a, b) => b.at - a.at);
}

/* How many of each decision, for the dialog's summary line. */
export function historySummary(events: PermissionEvent[]): string {
  if (!events.length) return "No tool asks decided in this chat yet.";
  const allowed = events.filter((e) => e.decision === "allow").length;
  const denied = events.length - allowed;
  const byRule = events.filter((e) => e.how === "rule").length;
  const parts = [`${allowed} allowed`, `${denied} denied`];
  if (byRule) parts.push(`${byRule} by rules`);
  return parts.join(" · ") + (events.length >= 200 ? " (the last 200)" : "");
}

File: chat/internal/tui/rules.go
package tui

import (
	"context"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Permission rules and history (docs/claude-parity.md, round 2 B): the
// workspace's and the chat's allow / deny / ask rules in Claude Code's
// own syntax (chats/rules.go), listed, added and removed with /rules;
// `A` on a tool ask (or /allow) remembers the call's rule for the whole
// workspace where `a` remembers it for the chat; /permissions lists how
// the chat's asks were decided and by whom.

// Rule mirrors chats.Rule.
type Rule struct {
	ID      string  `json:"id"`
	Kind    string  `json:"kind"`
	Pattern string  `json:"pattern"`
	Origin  string  `json:"origin"`
	ChatID  string  `json:"chatID"`
	By      *Actor  `json:"by"`
	At      float64 `json:"at"`
}

// Actor mirrors conversation.Actor: who added a rule or answered a card.
type Actor struct {
	PrincipalID string `json:"principalID"`
	Email       string `json:"email"`
	Name        string `json:"name"`
}

// Label is the person's name, email or principal.
func (a *Actor) Label() string {
	switch {
	case a == nil:
		return ""
	case a.Name != "":
		return a.Name
	case a.Email != "":
		return a.Email
	case a.PrincipalID == "owner":
		return "the owner"
	}
	return a.PrincipalID
}

// RulesView mirrors chats.RulesView: the workspace's rules and its chats'.
type RulesView struct {
	Workspace string      `json:"workspace"`
	Rules     []Rule      `json:"rules"`
	Chats     []ChatRules `json:"chats"`
}
type ChatRules struct {
	ID    string `json:"id"`
	Title string `json:"title"`
	Rules []Rule `json:"rules"`
}

// PermissionEvent mirrors chats.PermissionEvent: one decision on a tool
// ask.
type PermissionEvent struct {
	ID       string  `json:"id"`
	At       float64 `json:"at"`
	Tool     string  `json:"tool"`
	Summary  string  `json:"summary"`
	Decision string  `json:"decision"`
	How      string  `json:"how"`
	Rule     *Rule   `json:"rule"`
	Scope    string  `json:"scope"`
	By       *Actor  `json:"by"`
	Message  string  `json:"message"`
}

// Rules lists a chat's workspace's rules and its chats' own.
func (c *Client) Rules(ctx context.Context, chatID string) (RulesView, error) {
	var out RulesView
	err := c.do(ctx, "GET", "chats/"+chatID+"/rules", nil, &out)
	return out, err
}

// AddRule adds a rule to a workspace (sandbox id) or a chat (chat id).
func (c *Client) AddRule(ctx context.Context, kind, id, ruleKind, pattern string) (Rule, error) {
	var out Rule
	err := c.do(ctx, "POST", kind+"/"+id+"/rules", map[string]any{"kind": ruleKind, "pattern": pattern}, &out)
	return out, err
}

// RemoveRule removes a rule from a workspace or a chat; kind is
// "environments" or "chats".
func (c *Client) RemoveRule(ctx context.Context, kind, id, ruleID string) error {
	return c.do(ctx, "POST", kind+"/"+id+"/rules/"+ruleID+"/remove", map[string]any{}, nil)
}

// Permissions is a chat's permission history, oldest first.
func (c *Client) Permissions(ctx context.Context, chatID string) ([]PermissionEvent, error) {
	var out struct {
		Events []PermissionEvent `json:"events"`
	}
	err := c.do(ctx, "GET", "chats/"+chatID+"/permissions", nil, &out)
	return out.Events, err
}

// ruleRef is one numbered line of the last /rules listing: where the rule
// lives (environments or chats), the owner's id and the rule's.
type ruleRef struct {
	kind, owner, id string
}

// RulesListing lays the rules out numbered, the workspace's first then
// each chat's, and returns the references the numbers stand for.
func RulesListing(view RulesView, current string) (string, []ruleRef) {
	var b strings.Builder
	var refs []ruleRef
	line := func(r Rule) {
		refs = append(refs, ruleRef{})
		fmt.Fprintf(&b, "%2d  %-5s %-36s %s\n", len(refs), r.Kind, truncate(sanitize(r.Pattern), 36), sanitize(ruleOrigin(r, view)))
	}
	b.WriteString("workspace rules (every chat of this workspace)\n")
	if len(view.Rules) == 0 {
		b.WriteString("    none\n")
	}
	for _, r := range view.Rules {
		line(r)
		refs[len(refs)-1] = ruleRef{"environments", view.Workspace, r.ID}
	}
	for _, c := range view.Chats {
		if len(c.Rules) == 0 {
			continue
		}
		title := sanitize(c.Title)
		if c.ID == current {
			title += " (this chat)"
		}
		b.WriteString("chat " + title + "\n")
		for _, r := range c.Rules {
			line(r)
			refs[len(refs)-1] = ruleRef{"chats", c.ID, r.ID}
		}
	}
	b.WriteString("/rules add allow|deny|ask PATTERN adds a workspace rule (Bash(git *), Edit(src/**), Read, WebFetch(domain:x)) · /rules rm N removes one\ndeny answers without asking in every mode, ask makes a card even in auto, allow answers in ask mode")
	return b.String(), refs
}

// ruleOrigin says where a rule came from: the editor or an "allow always"
// answer, in which chat, by whom.
func ruleOrigin(r Rule, view RulesView) string {
	out := "from the editor"
	if r.Origin == "always" {
		out = "allow always"
		if r.ChatID != "" {
			for _, c := range view.Chats {
				if c.ID == r.ChatID {
					out += " in " + c.Title
				}
			}
		}
	}
	if by := r.By.Label(); by != "" {
		out += " by " + by
	}
	return out
}

// PermissionLines lays the history out, newest last, as the /permissions
// notice.
func PermissionLines(events []PermissionEvent, loc *time.Location) []string {
	if len(events) == 0 {
		return []string{"no tool asks decided in this chat yet"}
	}
	var out []string
	for _, ev := range events {
		when := time.Unix(int64(ev.At), 0).In(loc).Format("Jan 2 15:04")
		out = append(out, fmt.Sprintf("%s  %-5s %-12s %s  — %s", when, ev.Decision, truncate(ev.Tool, 12), truncate(sanitize(ev.Summary), 50), sanitize(permissionBy(ev))))
	}
	return out
}

// permissionBy says who or what decided an event.
func permissionBy(ev PermissionEvent) string {
	switch ev.How {
	case "auto":
		return "auto mode"
	case "rule":
		if ev.Rule != nil {
			return ev.Scope + " rule " + ev.Rule.Kind + " " + ev.Rule.Pattern
		}
		return "a rule"
	}
	by := ev.By.Label()
	if by == "" {
		by = "a card"
	}
	if ev.Rule != nil {
		by += ", always for the " + ev.Scope + " (" + ev.Rule.Pattern + ")"
	}
	if ev.Message != "" {
		by += ": " + ev.Message
	}
	return by
}

// rules runs /rules: the listing, `add KIND PATTERN` (a workspace rule)
// or `rm N` (a number from the listing).
func (a *App) rules(ctx context.Context, c *Chat, arg string) {
	if c == nil {
		a.setNotice("no chat selected")
		return
	}
	verb, rest, _ := strings.Cut(arg, " ")
	rest = strings.TrimSpace(rest)
	switch strings.ToLower(verb) {
	case "":
		view, err := a.Client.Rules(ctx, c.ID)
		if err != nil {
			a.setNotice(err.Error())
			return
		}
		text, refs := RulesListing(view, c.ID)
		a.ruleRefs = refs
		a.setNotice(text)
	case "add":
		kind, pattern, _ := strings.Cut(rest, " ")
		pattern = strings.Trim(strings.TrimSpace(pattern), "\"'")
		if kind == "" || pattern == "" {
			a.setNotice("/rules add allow|deny|ask PATTERN, such as /rules add deny \"Bash(rm *)\"")
			return
		}
		view, err := a.Client.Rules(ctx, c.ID)
		if err != nil {
			a.setNotice(err.Error())
			return
		}
		r, err := a.Client.AddRule(ctx, "environments", view.Workspace, kind, pattern)
		if err != nil {
			a.setNotice(err.Error())
			return
		}
		a.ruleRefs = nil
		a.setNotice("workspace rule added: " + r.Kind + " " + sanitize(r.Pattern))
	case "rm", "remove", "del":
		if a.ruleRefs == nil {
			view, err := a.Client.Rules(ctx, c.ID)
			if err != nil {
				a.setNotice(err.Error())
				return
			}
			_, a.ruleRefs = RulesListing(view, c.ID)
		}
		n, err := strconv.Atoi(rest)
		if err != nil || n < 1 || n > len(a.ruleRefs) {
			a.setNotice("/rules rm N with N from /rules")
			return
		}
		ref := a.ruleRefs[n-1]
		if err := a.Client.RemoveRule(ctx, ref.kind, ref.owner, ref.id); err != nil {
			a.setNotice(err.Error())
			return
		}
		a.ruleRefs = nil
		a.setNotice(fmt.Sprintf("rule %d removed", n))
	default:
		a.setNotice("/rules · /rules add allow|deny|ask PATTERN · /rules rm N")
	}
}

// permissions runs /permissions: the chat's decided asks.
func (a *App) permissions(ctx context.Context, c *Chat) {
	if c == nil {
		a.setNotice("no chat selected")
		return
	}
	events, err := a.Client.Permissions(ctx, c.ID)
	if err != nil {
		a.setNotice(err.Error())
		return
	}
	a.setNotice(strings.Join(PermissionLines(events, a.now().Location()), "\n"))
}

// allowAlways runs /allow [chat]: the first pending tool ask allowed
// always for the workspace (or the chat).
func (a *App) allowAlways(ctx context.Context, c *Chat, arg string) {
	if c == nil {
		a.setNotice("no chat selected")
		return
	}
	scope := "workspace"
	if strings.EqualFold(arg, "chat") {
		scope = "chat"
	}
	for _, ap := range c.Pending() {
		if p := ap.Permission(); p != nil && !p.IsPlan() {
			a.answerAlways(ctx, c.ID, ap, p, scope)
			return
		}
	}
	a.setNotice("no tool ask pending")
}

// answerAlways allows an ask and remembers its rule at the scope.
func (a *App) answerAlways(ctx context.Context, chatID string, ap Approval, p *Permission, scope string) {
	if err := a.Client.Answer(ctx, chatID, ap.ID, true, true, scope, "", ""); err != nil {
		a.setNotice(err.Error())
		return
	}
	a.setNotice("allowed always for this " + scope + ": " + sanitize(p.Always))
}
